sources/web/datalab/idleTimeout.ts

```diff
@@ -0,0 +1,256 @@
+  }
+}
+
+function shutdownIfTimedOut() {
+  if (hasTimedOut()) {
+    // TODO - check to see if the kernel is idle; if not, don't try to stop the VM
+    try {
+      logging.getLogger().debug('Idle timeout, shutting down with this command: ' + shutdownCommand);
+      childProcess.execSync(shutdownCommand, {env: process.env});
+      logging.getLogger().debug('Shutdown command succeeded');
+    } catch (err) {
+      logging.getLogger().error(err, 'Shutdown failed. stderr: %s', err.stderr);
+    }
+  }
+}
+
+export function startChecker() {
+  setInterval(shutdownIfTimedOut, idleCheckIntervalSeconds*1000);
+}
+
+/**
+ * Implements timeout management request handling.
+ * @param request the incoming http request.
+ * @param response the outgoing http response.
+ */
+function requestHandler(request: http.ServerRequest, response: http.ServerResponse): void {
+  const userId = userManager.getUserId(request);
+  if ('POST' == request.method) {
+    postTimeoutHandler(userId, request, response);
+  } else {
+    getTimeoutHandler(userId, request, response);
+  }
+}
+
+/**
+ * Handles 'GET' requests to the timeout management handler.
+ * @param request the incoming http request.
+ * @param response the outgoing http response.
+ */
+function getTimeoutHandler(userId: string, request: http.ServerRequest, response: http.ServerResponse): void {
+  const timeoutInfo:any = {
+    enabled: isTimeoutEnabled(),
+    secondsRemaining: timeoutSecondsRemaining(),
+    idleTimeoutSeconds: idleTimeoutSeconds
+  };
+  const parsedUrl = url.parse(request.url, true);
+  if ('key' in parsedUrl.query) {
+    const key = parsedUrl.query['key'];
+    if (key in timeoutInfo) {
+      response.writeHead(200, { 'Content-Type': 'application/json' });
+      response.end(JSON.stringify(timeoutInfo[key]));
+    } else {
+      response.writeHead(404, { 'Content-Type': 'text/plain' });
+      response.end();
+    }
+  } else {
+    response.writeHead(200, { 'Content-Type': 'application/json' });
+    response.end(JSON.stringify(timeoutInfo));
+  }
+}
+
+/**
+ * Handles 'POST' requests to the timeout management handler.
+ * @param request the incoming http request.
+ * @param response the outgoing http response.
+ */
+function postTimeoutHandler(userId: string, request: http.ServerRequest, response: http.ServerResponse): void {
+  let formData : any;
+  let body : string = "";
+  request.on('data', function(chunk: string) { body += chunk; });
+  request.on('end', function() {
+    if (body) {
+      formData = querystring.parse(body);
+    } else {
+      const parsedUrl = url.parse(request.url, true);
+      formData = parsedUrl.query;
+    }
+    logging.getLogger().debug('got form data, type=' + typeof(formData));
+    logging.getLogger().debug(formData);
+    if (formData.enabled) {
+      const newEnabled = formData.enabled == 'true' ? true : false;
+      idleTimeoutEnabled = newEnabled;
+      reset();  // Reset the timeout value when timeout is enabled or disabled
```

Comment:
looks like this and the if statement below are redundant, since if a request is handled here it means it'll reset anyway.